### ISSUE TITLE
Improve situation about incompatible versions

### ISSUE THREAD

**Issue by pepijndevos** [rhempel/ev3dev-lang-python#277]
Looking at the issues on this project and my own experience over the past few days, a lot of problems stem from having incompatible versions of python, linux and this library.

From a fresh and unmodified ev3dev image, I was first trapped in Python 2 with a completely broken library, and then with an outdated library in Python 3 which broke in more subtle ways.

I don't get why the kernel has to have breaking changes all the time, and I don't get why you'd ship an image with broken libraries in it.

But that being the case (and possibly outside your control), it would be nice if the library would warn me if the kernel version is newer than the library version.

These are tough problems that I don't have an immediate answer to (or I'd have made a PR), but I'd at least like to open the discussion for how to handle this, as it can greatly improve the experience of new users. Ideally everyone should just update their software before doing anything else, but on a platform where doing so can take a few hours, this is a bit much to expect.

**Comment by WasabiFan** [comment]
> Looking at the issues on this project and my own experience over the past few days, a lot of problems stem from having incompatible versions of python, linux and this library.

Keep in mind that you are looking at the _open_ issues, which is only 18 relative to the hundred or so that have been resolved. We've had issues stemming from the fact that the ev3dev image was so old, but that has since been resolved by the new image that was released last month. My guess is that most of the open issues are continual workstreams or discussions as opposed to acute unadressed issues.

> From a fresh and unmodified ev3dev image, I was first trapped in Python 2 with a completely broken library, and then with an outdated library in Python 3 which broke in more subtle ways.

What image? This shouldn't be the case in the November image.

> I don't get why the kernel has to have breaking changes all the time, and I don't get why you'd ship an image with broken libraries in it.

The last breaking changes were at the beginning of this year, I think -- roughly in March. We've been releasing snapshot builds since then which included those changes, and last month released an official image with those changes included. I'm unhappy that it took so long, but we now have what should be a fully-functional image available as an official release. The README should also be up-to-date.

> But that being the case (and possibly outside your control), it would be nice if the library would warn me if the kernel version is newer than the library version.

@ddemidov Thoughts? This makes sense to me, however it's difficult to do this correctly. Most kernel upgrades _don't_ require library changes and aren't accompanied by a library release, so you can't outright prevent execution on a newer platform version, and even printing a warning could get annoying. In general we don't expect many breaking changes, especially as ev3dev approaches 1.0.

**Comment by ddemidov** [comment]
We could utilize [supportedKernel](https://github.com/ev3dev/ev3dev-lang/blob/b34680f210de2bb49edf2b5eee93ae5b8c264c8d/spec.json#L5) field from `spec.json` together with [os.uname()](https://docs.python.org/3/library/os.html#os.uname) function to check if the combination of library and kernel is good. Although we would probably have to convert that value to a range/list. Also, updating the section in `spec.json` could quickly become annoying, easy to forget, source of frustrating warnings/errors when we do forget to update the spec or regenerate the library.

So I think the only thing we should do is to make sure the python2 version of the library is not installed in the new images.

**Comment by WasabiFan** [comment]
> So I think the only thing we should do is to make sure the python2 version of the library is not installed in the new images.

Agreed. I believe that is just dependent on the OpenRoberta connector getting upgraded to Python 3.

**Comment by pepijndevos** [comment]
Iirc my image is from October(latest stable from a few weeks back), but I'll get back with actual versions and dates tomorrow.As of last night, after upgrading the library via apt, it still thinks it should ‎write stop_command while the kernel only knows about stop_action. I'm sure this means either or both are still horribly outdated.I ran a full dist-upgrade overnight, so we'll see if that helped. Else I'll try messing with pip or try a newer/nightly image.Suffices to say, I've already wasted quite a few hours to version incompatibilities.But I agree that if the kernel will reach stable version soon, probably the best is to delete all broken stuff and make sure the rest keeps working for eternity.

**Comment by ddemidov** [comment]
Are you sure you are calling the scripts with python3? There is a deprecated python2 version of the library that is currently getting pulled as a dependency of openroberta.

**Comment by WasabiFan** [comment]
Are you running your programs with `python3` as opposed to `python`? You need to if you want to get the correct version of the libraries. No upgrading or configuration should be required.

**Comment by pepijndevos** [comment]
Pretty sure, as the Python 2 version fails with a _path attribute error. The Python 3 version runs the motor, but just coasts them. I'll check which version I have and where it came from tomorrow. ‎Also note that no warning is given for setting unknown keyword arguments. It just silently set stop_action on the object without doing anything to the motor. Would you accept a pr that warns/errors on that? Prevents bugs and typos IMO, but more dynamic folks might call it a feature.

**Comment by ddemidov** [comment]
`stop_action` is the correct property name, so I assume you meant `stop_command`. The library does allow you to set `stop_command`, as in
```py
from ev3dev.ev3 import *
m = LargeMotor()
m.stop_command = 'brake'
```
but that is just a 'feature' of python: here we are creating a new field called `stop_command` in the `m` instance. I guess you could check that keyword arguments in `run_*` commands belong to a fixed set, but that would introduce slight overhead for everyone (may be not that important, since we are using python anyway), and also could break with possible introduction of new attributes.

**Comment by pepijndevos** [comment]
I have the opposite problem: If I set stop_action it just sets the property, if I set stop_command I get an IO error. So it's the library that's outdated. dir(LargeMotor) also shows a stop_commands that gives an IO error when reading.

**Comment by ddemidov** [comment]
What does `uname -a` gives you (when entered into the command prompt of an ssh session)?
Also, how exactly are you launching your scripts? Can you paste a complete but simple example here?

**Comment by pepijndevos** [comment]
It seems the full dist-upgrade, a reboot and some sleep resolved my particular issue. I just ran `ipython3` and ran a few lines from history. Everything that was broken yesterday works now. For completeness, I'm now on version 0.8.0 on kernel `Linux ev3dev 4.4.32-17-ev3dev-ev3 #1 PREEMPT Wed Nov 16 11:12:06 CST 2016 armv5tejl GNU/Linux`.

Here is the apt log: https://gist.github.com/anonymous/70f0a70bef41ed803f2cd8694493a732
After upgrading just the library it did not seem to work yet, but maybe some caching was in play, as I was still seeing `LargeMotor.stop_commands` as an attribute.

**Comment by WasabiFan** [comment]
Glad you got it working! I'm still confused, though; it should "just work" with the standard image. What image did you flash with? Can you paste the output of `ev3dev-sysinfo -m`? I'd like to get some level of understanding here so that we know what caused this. You shouldn't have had to go through all these extra steps.

**Comment by pepijndevos** [comment]
```
Image file:         ev3dev-jessie-ev3-generic-2016-10-17
Kernel version:     4.4.32-17-ev3dev-ev3
Board:              LEGO MINDSTORMS EV3 Programmable Brick
Revision:           0006
Brickman:           0.8.0
ev3devKit:          0.4.2
```

**Comment by WasabiFan** [comment]
Yeah, that seems right. The image file is the correct one and so is the kernel version. Given that I was able to flash that same image and immediately run working Python motor code earlier, I'm thinking that this was an issue with test procedure, caching, typos or similar. Any other thoughts?

**Comment by pepijndevos** [comment]
`¯\_(ツ)_/¯`

Can I uninstall the Python 2 version if I'm not using the Roberto thing? Even while knowing it's broken, just by having it, being sleepy and forgetting a 3 can cause hours of agony, it turns out.

**Comment by ddemidov** [comment]
You could just put a shebang line into your scripts:
```py
#!/usr/bin/env python3
import ev3dev.ev3 as ev3
...
```
, make the script executable (`chmod +x script.py`), then launch it with `./script.py` or directly from brickman. With that, you don't have to remember to use the correct interpreter anymore.

**Comment by pepijndevos** [comment]
About setting unknown attributes: I was thinking that if you declare `__slots__` for all classes, this is not slower, and even saves memory.

**Comment by ddemidov** [comment]
I guess that could work. Do you want to create a pull request? You would have to keep in mind that the code is partially [generated from specification](https://github.com/ev3dev/ev3dev-lang/tree/develop/autogen), ([example](https://github.com/rhempel/ev3dev-lang-python/blob/488b89f169c685be26799ae9b73df99f74c6a89f/ev3dev/core.py#L1487-L1576)) but some methods/properties are added by hand ([example](https://github.com/rhempel/ev3dev-lang-python/blob/488b89f169c685be26799ae9b73df99f74c6a89f/ev3dev/core.py#L1597-L1612)), so the declaration of `__spec__` would have to allow merging generated and manual lists together.

**Comment by pepijndevos** [comment]
So I have to create a PR on both the autogen and this lib with the generated code? I'll look into it during the holiday break.

**Comment by ddemidov** [comment]
PR here should be enough, as liquid templates used to generate the code are stored locally, in the [templates](https://github.com/rhempel/ev3dev-lang-python/tree/develop/templates) folder.

**Comment by ensonic** [comment]
@WasabiFan , @ddemidov : openroberta has been converted to python3 already, it should not pull in python2 - I double checked my debina/control file.

**Comment by WasabiFan** [comment]
Does the October image contain the new version?

**Comment by pepijndevos** [comment]
@WasabiFan probably not. After my upgrade, apt now says 
```
The following packages were automatically installed and are no longer required:
  python-bluez python-dbus python-dbus-dev python-ev3dev python-pil
Use 'apt-get autoremove' to remove them.
```

**Comment by ddemidov** [comment]
I tend to do most of the programming on my PC. When I want to test if something works, I commit to my local branch on the PC, then pull from the brick from that branch (over ssh).
When you are on a brick, its enough to `cd` into `ev3dev-lang-python` for `import ev3dev.ev3` to import from the local copy that is being developed/tested, and not from the system-wide package.

**Comment by ddemidov** [comment]
> After my upgrade, apt now says

So may be we should ask @dlech to remove those packages from the image.

**Comment by pepijndevos** [comment]
I'm obviously not using the right unittest incantations
`Ran 0 tests in 0.093s`

**Comment by ddemidov** [comment]
```
$ git clone --recursive https://github.com/rhempel/ev3dev-lang-python
...
$ cd ./ev3dev-lang-python/tests/
$ ./api_tests.py 
...
Ran 3 tests in 0.022s
```

**Comment by dlech** [comment]
> Does the October image contain the new version?

no

> So may be we should ask @dlech to remove those packages from the image.

I can't change an image that has already been released. :wink:. Use the latest [snapshot](https://oss.jfrog.org/list/oss-snapshot-local/org/ev3dev/brickstrap/) if you want an image that includes the most up-to-date packages out of the box.

**Comment by ddemidov** [comment]
The problem is the image contains packages that should not be there (python2 version of the python library).

**Comment by WasabiFan** [comment]
The python 2 version is included as a dependency of the old OpenRoberta connector. Any image built now would contain the python 3 version of the OpenRoberta connector and by extension not depend on the old version of this library.

**Comment by ddemidov** [comment]
I think I misinterpreted the `apt-get` output @pepijndevos  [pasted above](https://github.com/rhempel/ev3dev-lang-python/issues/277#issuecomment-267606257). That is probably after `apt-get upgrade`, and I thought that was the case with a freshly flashed image.

**Comment by pepijndevos** [comment]
Correct, sorry for the confusion.

**Comment by ensonic** [comment]
Also to conclude we probably need more vetting on images, at least w.r.t. our own components. I could have release a python3 version of the openroberta package if'd had known that the october image is planned.

**Comment by WasabiFan** [comment]
@ensonic I'm fairly sure that we pointed out that we were planning on releasing an image, going as far as tagging you and saying that OpenRoberta was blocking release. Looking at past conversations, we knew we were going to be releasing an image as early as September and during that month we released this library in preparation. OpenRoberta was mentioned multiple times during those conversations. Wasn't that why you began transitioning to Python 3?
